pkg/operator/targetconfigcontroller/targetconfigcontroller.go

```diff
@@ -412,6 +445,113 @@ func ensureLocalhostRecoverySAToken(ctx context.Context, client coreclientv1.Cor
 	return err
 }
 
+func ensureMachineKubeconfigs(ctx context.Context, client coreclientv1.CoreV1Interface, infrastructureLister configv1.InfrastructureLister, recorder events.Recorder) error {
+	requiredSecret := resourceread.ReadSecretV1OrDie(v410_00_assets.MustAsset("v4.1.0/kube-apiserver/machine-kubeconfigs.yaml"))
+
+	systemAdminCredsSecret, err := client.Secrets(operatorclient.OperatorNamespace).Get(ctx, "machine-system-admin-client", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	systemAdminClientCert := systemAdminCredsSecret.Data[corev1.TLSCertKey]
+	if len(systemAdminClientCert) == 0 {
+		return fmt.Errorf("missing system:admin client certificate")
+	}
+	systemAdminClientKey := systemAdminCredsSecret.Data[corev1.TLSPrivateKeyKey]
+	if len(systemAdminClientKey) == 0 {
+		return fmt.Errorf("missing system:admin client private key")
+	}
+
+	servingCABundleCM, err := client.ConfigMaps(operatorclient.TargetNamespace).Get(ctx, "kube-apiserver-server-ca", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	servingCABundleData := servingCABundleCM.Data["ca-bundle.crt"]
+	if len(servingCABundleData) == 0 {
+		return fmt.Errorf("missing serving CA bundle data")
+	}
+
+	infrastructure, err := infrastructureLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+	apiServerInternalURL := infrastructure.Status.APIServerInternalURL
+	if len(apiServerInternalURL) == 0 {
+		return fmt.Errorf("missing APIServerInternalURL")
+	}
+	apiServerURL := infrastructure.Status.APIServerURL
+	if len(apiServerURL) == 0 {
+		return fmt.Errorf("missing APIServerURL")
+	}
+
+	for kubeconfigName, kubeconfigBytes := range requiredSecret.Data {
+		obj, err := runtime.Decode(clientcmdDecoder, kubeconfigBytes)
```

Comment:
decoding and re-encoding should be unnecessary. Seems like you could simply string replace.